javatests/arcs/sdk/HandleHolderBaseTest.kt

```diff
@@ -0,0 +1,303 @@
+    verify(writeMock, times(1)).unregisterForStorageEvents()
+  }
+
+  @Test
+  fun reset_allHandles_getUnregistered() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    val readMock = mock<Handle>()
+    val writeMock = mock<Handle>()
+    handleHolder.setHandle("readHandle", readMock)
+    handleHolder.setHandle("writeHandle", writeMock)
+
+    handleHolder.reset()
+
+    verify(readMock, times(1)).unregisterForStorageEvents()
+    verify(writeMock, times(1)).unregisterForStorageEvents()
+  }
+
+  @Test
+  fun reset_allHandles_getClosed() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    val readMock = mock<Handle>()
+    val writeMock = mock<Handle>()
+    handleHolder.setHandle("readHandle", readMock)
+    handleHolder.setHandle("writeHandle", writeMock)
+
+    handleHolder.reset()
+
+    verify(readMock, times(1)).close()
+    verify(writeMock, times(1)).close()
+  }
+
+  @Test
+  fun reset_allHandles_allElementsRemoved() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf(
+        "readHandle" to mock(),
+        "writeHandle" to mock()
+      )
+    )
+    handleHolder.setHandle("readHandle", mock())
+    handleHolder.setHandle("writeHandle", mock())
+    assertThat(handleHolder.handles).isNotEmpty()
+
+    handleHolder.reset()
+
+    assertThat(handleHolder.handles).isEmpty()
+  }
+
+  @Test
+  fun isEmpty_delegatesToHandlesMap_returnsTrue() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    )
+
+    assertThat(handleHolder.isEmpty()).isTrue()
+  }
+
+  @Test
+  fun isEmpty_delegatesToHandlesMap_returnsFalse() {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    )
+
+    handleHolder.setHandle("readHandle", mock())
+
+    assertThat(handleHolder.isEmpty()).isFalse()
+  }
+
+  @Test
+  fun createForeignReference_noHandles_asserts() = runBlockingTest {
+    val handleHolder = HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    )
+
+    assertFailsWith<IllegalStateException> {
+      handleHolder.createForeignReference<EntityBase>(mock(), "someId")
+    }
+  }
+
+  @Test
+  fun createForeignReference_someHandles_success() = runBlockingTest {
+    class DummyHandle : HandleHolderBase(
+      "TestParticle",
+      entitySpecs = mapOf("readHandle" to mock())
+    ) {
+      val readHandle: Handle by handles
+    }
+
+    val handleHolder = DummyHandle()
+    val mock = mock<Handle>()
+    handleHolder.setHandle("readHandle", mock)
+
+    handleHolder.readHandle.createForeignReference<EntityBase>(mock(), "someId")
```

Comment:
- sorry, i'm not sure i understand - why do you need the special DummyHandle class, instead of initializing the HandleHolderBase like in other tests?
- doesn't calling `handleHolder.readHandle.createForeignReference` actually tests the handle's createForeignReference call? why not `handleHolder.createForeignReference`?

Reply:
Yes, thank you. The former was needed to do the latter, which as you pointed out, isn't correct.